hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java

```diff
@@ -0,0 +1,770 @@
+import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.codec.CacheAddInvalidationListenerCodec;
+import com.hazelcast.client.impl.protocol.codec.CacheAddNearCacheInvalidationListenerCodec;
+import com.hazelcast.client.impl.protocol.codec.CacheRemoveEntryListenerCodec;
+import com.hazelcast.client.spi.ClientClusterService;
+import com.hazelcast.client.spi.ClientContext;
+import com.hazelcast.client.spi.ClientListenerService;
+import com.hazelcast.client.spi.EventHandler;
+import com.hazelcast.client.spi.impl.ClientInvocationFuture;
+import com.hazelcast.client.spi.impl.ListenerMessageCodec;
+import com.hazelcast.client.util.ClientDelegatingFuture;
+import com.hazelcast.config.CacheConfig;
+import com.hazelcast.config.InMemoryFormat;
+import com.hazelcast.config.NativeMemoryConfig;
+import com.hazelcast.config.NearCacheConfig;
+import com.hazelcast.core.ExecutionCallback;
+import com.hazelcast.internal.adapter.ICacheDataStructureAdapter;
+import com.hazelcast.internal.nearcache.NearCache;
+import com.hazelcast.internal.nearcache.NearCacheManager;
+import com.hazelcast.internal.nearcache.impl.invalidation.RepairingHandler;
+import com.hazelcast.internal.nearcache.impl.invalidation.RepairingTask;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.Connection;
+import com.hazelcast.nio.serialization.Data;
+
+import javax.cache.expiry.ExpiryPolicy;
+import javax.cache.integration.CompletionListener;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static com.hazelcast.cache.impl.ICacheService.SERVICE_NAME;
+import static com.hazelcast.config.InMemoryFormat.NATIVE;
+import static com.hazelcast.config.NearCacheConfig.LocalUpdatePolicy.CACHE;
+import static com.hazelcast.config.NearCacheConfig.LocalUpdatePolicy.CACHE_ON_UPDATE;
+import static com.hazelcast.instance.BuildInfo.UNKNOWN_HAZELCAST_VERSION;
+import static com.hazelcast.instance.BuildInfo.calculateVersion;
+import static com.hazelcast.internal.nearcache.NearCache.CACHED_AS_NULL;
+import static com.hazelcast.internal.nearcache.NearCache.NOT_CACHED;
+import static com.hazelcast.internal.nearcache.NearCacheRecord.NOT_RESERVED;
+import static com.hazelcast.util.ExceptionUtil.rethrow;
+import static com.hazelcast.util.MapUtil.createHashMap;
+import static com.hazelcast.util.Preconditions.checkTrue;
+import static java.lang.String.format;
+
+/**
+ * An {@link ICacheInternal} implementation which handles near cache specific behaviour of methods.
+ *
+ * @param <K> the type of key.
+ * @param <V> the type of value.
+ */
+@SuppressWarnings({"checkstyle:methodcount", "checkstyle:classdataabstractioncoupling",
+        "checkstyle:classfanoutcomplexity", "checkstyle:anoninnerlength"})
+public class NearCachedClientCacheProxy<K, V> extends ClientCacheProxy {
```

Comment:
This class seems to have a code coverage of 64% percent (if I didn't miss any important tests). We should add the missing tests as soon as possible.

It looks like the whole `getAsyncInternal()`, `GetAsyncCallback`, `onPutIfAbsentAsyncInternal()`, `onFailureInternal()`, `onRemoveAsyncInternal()`, `onGetAndRemoveAsyncInternal()`, `onReplaceInternalAsync()`, `onReplaceAndGetAsync()`, `CacheOrInvalidateCallback`, `removeAllKeysInternal()`, `removeAll()` and ` invokeInternal()` is untested.

If the public API methods who call these are also available for `IMap` etc. we directly should add those tests to the unified Near Cache tests, so the other data structures are also tested. This might be out of the scope of this PR though.